src/itdelatrisu/opsu/translation/I18n.java

```diff
@@ -0,0 +1,346 @@
+	 * @param fallbackId A fallback ID to identify this locale if the key <code>language.name</code> is not present.
+	 */
+	private I18n(InputStream stream, String fallbackId) {
+		this.keys = new HashMap<String, String>();
+		if (stream == null)
+			return; // used by the hardcoded fallback translator
+
+		try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8))) {
+			String line;
+			while ((line = reader.readLine()) != null) 
+				if (!line.isEmpty() && line.charAt(0) != 35) {
+					String[] values = line.split("=", 2);
+					this.keys.put(values[0], values[1]);
+				}
+			
+		} catch (IOException e) {
+			Log.warn("Could not load all translation keys", e);
+		}
+
+		if (!keys.containsKey("language.name"))
+			keys.put("language.name", fallbackId);
+	}
+
+	/**
+	 * Constructor for language files that can be found outside the JAR
+	 * @param langFile The reference to the external language file
+	 */
+	private I18n(File langFile) throws FileNotFoundException {
+		this(new FileInputStream(langFile), langFile.getName().substring(0, langFile.getName().length() - 5));
+	}
+
+	/**
+	 * Constructor for language files found inside the JAR
+	 * @param resourceLocation A textual representation of the location of the
+	 * JAR resource
+	 */
+	private I18n(String resourceLocation) {
+		this(I18n.class.getResourceAsStream(resourceLocation),
+				resourceLocation.substring(resourceLocation.lastIndexOf('/'), resourceLocation.length() - 5));
+	}
+
+	/**
+	 * <p>
+	 * Searches for and loads all locales found. Calling this method again will
+	 * clear all active locales and reload them from disk.
+	 * </p><p>
+	 * Upon calling this method on reload, it is imperative to call any of the
+	 * <code>setLocale()</code> commands for the changes to take effect.
+	 * </p>
+	 * @see #setLocale(String)
+	 * @see #setLocale(int)
+	 */
+	public static void load() {
+		if (!LOCALES.isEmpty() || localeIDs != null) {
+			LOCALES.clear();
+			localeIDs = null;
+			currentLocale = null;
+			Utils.gc(false);
+		}
+
+		final Pattern fileExt = Pattern.compile("(.+).(lang)$");
+		currentLocale = FALLBACK;
+
+		try (JarFile opsu = Utils.getJarFile()) {
+			Enumeration<JarEntry> entries = opsu.entries();
+			while (entries.hasMoreElements()) {
+				JarEntry entry = entries.nextElement();
+				String entryName = entry.getName();
+				if (fileExt.matcher(entryName).matches())
+					addLocale(new I18n(entryName));
+
+			}
+		} catch (NullPointerException e) {
+			// compensate for running outside of JAR
+			for (File langFile : findRecursively(Utils.getRunningDirectory(), fileExt)) 
+				try {
+					addLocale(new I18n(langFile));
+				} catch (Exception e2) {
+					Log.error("Could not load translation: " + langFile.getName(), e2);
+				}
+			
+		} catch (IOException e) {
+			Log.error("Could not find resources for translation", e);
+		}
+
+		localeIDs = new String[LOCALES.size()];
```

Comment:
Consider using some sort of [`List`](https://docs.oracle.com/javase/7/docs/api/java/util/List.html) (maybe an [`ArrayList`](https://docs.oracle.com/javase/7/docs/api/java/util/ArrayList.html)). It'll also simplify the `for` loop below a little bit. If `localeIDs` and `LOCALES` are tightly coupled, you may want to consider using a [`Map`](https://docs.oracle.com/javase/7/docs/api/java/util/Map.html) from IDs to locales instead.